packages/task/src/browser/task-terminal-widget-manager.ts

```diff
@@ -0,0 +1,184 @@
+import { TaskDefinitionRegistry } from './task-definition-registry';
+import { WorkspaceService } from '@theia/workspace/lib/browser/workspace-service';
+
+export interface TaskTerminalWidgetOpenerOptions extends WidgetOpenerOptions, TerminalWidgetFactoryOptions {
+    taskId: number;
+    taskConfig?: TaskConfiguration;
+}
+
+export interface TaskTerminalWidget extends TerminalWidget {
+    readonly kind: 'task';
+    dedicated?: boolean;
+    taskId?: number;
+    taskConfig?: TaskConfiguration;
+    busy?: boolean;
+}
+export namespace TaskTerminalWidget {
+    export function is(widget: TerminalWidget): widget is TaskTerminalWidget {
+        return widget.kind === 'task';
+    }
+}
+
+export namespace TaskTerminalWidgetOpenerOptions {
+    export function isDedicatedTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return !!options.taskConfig && !!options.taskConfig.presentation && options.taskConfig.presentation.panel === PanelKind.Dedicated;
+    }
+
+    export function isNewTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return !!options.taskConfig && !!options.taskConfig.presentation && options.taskConfig.presentation.panel === PanelKind.New;
+    }
+
+    export function isSharedTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return !!options.taskConfig &&
+            (options.taskConfig.presentation === undefined || options.taskConfig.presentation.panel === undefined || options.taskConfig.presentation.panel === PanelKind.Shared);
+    }
+}
+
+@injectable()
+export class TaskTerminalWidgetManager {
+
+    @inject(ApplicationShell)
+    protected readonly shell: ApplicationShell;
+
+    @inject(TaskDefinitionRegistry)
+    protected readonly taskDefinitionRegistry: TaskDefinitionRegistry;
+
+    @inject(TerminalService)
+    protected readonly terminalService: TerminalService;
+
+    @inject(TaskWatcher)
+    protected readonly taskWatcher: TaskWatcher;
+
+    @inject(TaskServer)
+    protected readonly taskServer: TaskServer;
+
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+
+    @postConstruct()
+    protected init(): void {
+        this.taskWatcher.onTaskExit((event: TaskExitedEvent) => {
+            const finishedTaskId = event.taskId;
+            // find the terminal where the task ran, and mark it as "idle"
+            for (const widget of this.getTaskTerminalWidgets()) {
+                if (widget.taskId === finishedTaskId) {
+                    widget.busy = false;
+                    widget.scrollToBottom();
+                    widget.writeLine('\x1b[1m\n\rTerminal will be reused by tasks. \x1b[0m\n');
+                    break;
+                }
+            }
+        });
+
+        const context = this.workspaceService.workspace && this.workspaceService.workspace.uri;
+        this.terminalService.onDidCreateTerminal((widget: TerminalWidget) => {
```

Comment:
I think it should be properly guarded to be type-safe and don't mutate all terminals:
```ts
this.terminalService.onDidCreateTerminal(widget => {
    const terminal = TaskTerminalWidget.is(widget) && widget;
    if (terminal) {
        const didConectListener = terminal.onDidConnect(async () => {
        const tasksInfo = await this.taskServer.getTasks(context);
        const taskInfo = tasksInfo.find(info => info.terminalId === widget.terminalId);
        if (taskInfo) {
            const taskConfig = taskInfo.config;
            terminal.dedicated = !!taskConfig.presentation && !!taskConfig.presentation.panel && taskConfig.presentation.panel === PanelKind.Dedicated;
            terminal.taskId = taskInfo.taskId;
            terminal.taskConfig = taskConfig;
            terminal.busy = true;
        }
    });
    widget.onDidDispose(() => didConectListener.dispose());
}
```